packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts

```diff
@@ -170,11 +195,19 @@ const getFilterCallbackFromItem = (
     return null;
   }
 
-  const fn = (rowId: GridRowId) => {
-    const cellParams = apiRef.current.getCellParams(rowId, newFilterItem.field!);
+  const isFastFilter = isFastFilterFn(applyFilterOnRow);
 
-    return applyFilterOnRow(cellParams);
-  };
+  const fn =
+    isFastFilter ?
+      (rowId: GridRowId) => {
+        const value = apiRef.current.getCellValue(rowId, newFilterItem.field!);
+        const params = { value } as GridCellParams<any, unknown, unknown, GridTreeNode>;
+        return applyFilterOnRow(params);
+      } :
+      (rowId: GridRowId) => {
+        const params = apiRef.current.getCellParams(rowId, newFilterItem.field!);
+        return applyFilterOnRow(params);
+      };
```

Comment:
Another option might be using [lazy getters](https://www.merrickchristensen.com/articles/lazy-evaluation-in-javascript/) for `cellParams` properties.
The benefit is that it will also make custom filters faster (as long as they only use `value`).
What do you think?

Reply:
It could be. It would still require a `defineProperty` call, which is more expensive than raw props (it's the reason `_objectSpread2` is so slow, it uses `defineProperties`).

IMHO I think we should introduce a new filtering function API that would look like `(value: any, id: GridRowId, field: string) => boolean`. This new signature is efficient, minimal, and we could add the `apiRef` as an escape hatch in case the function needs to call more expensive functions like `.getCellParams`. We don't allocate any object with that signature. We could migrate our internal filters to this new API, and keep the old API for external filters until we do a major release.

Reply:
I've removed the "fast filters" concept in favor of new "V7" APIs.

Reply:
So the plan is:

v6:
- introduce `getApplyFilterFnV7`
- recommend using `getApplyFilterFnV7` in the Performance docs
- deprecate `getApplyFilterFn`?

v7:
- remove `getApplyFilterFn`
- rename `getApplyFilterFnV7` to `getApplyFilterFn`
- add migration guide for the new `getApplyFilterFn` signature

Is this correct?

Reply:
Yes, that would be the plan. There are other places where we could also apply the same strategy (in other PRs, this is big already).

Reply:
We could also just introduce a new name, e.g. `getFilterFn`, and keep `getApplyFilterFn` working for the forseeable future. Depends how averse we are to breaking changes in major releases. Forcing a breaking change in v7 however gives the chance to users to improve their performance, which they won't if things keep working.

Reply:
Ah I see. Maybe we do this:
 - We define a wrapper-creator that turns `getApplyFilterFnV7` into `getApplyFilterFn`: `const createLegacyFilter = (filterFn) => ({ value, row, colDef }, apiRef) => filterFn(value, row, colDef, apiRef)`
 - We make all our internal functions V7, but we use the wrapper-creator to auto-create legacy `.getApplyFilterFn` functions on our exported operators to avoid breaking the API.
 - We accept both `.getApplyFilterFn` and `.getApplyFilterFnV7` but we prioritize the V7 one if present.

Reply:
OTOH we could also add a `.v7` flag on the function object like we do for `createSelector` output selectors. I think this would simplify the migration, no need for `V7` APIs, we just add the new supported signature to the API (`getApplyFilterFn: LegacyFilterFn | V7FilterFn`). It would also make the performance improvements available right away to users re-using our exported `.getApplyFilterFn`.

I don't have a strong opinion on which path is better, let me know what you think.

Reply:
> OTOH we could also add a .v7 flag on the function object like we do for createSelector output selectors

I like this idea, it seems simpler 👍

Reply:
For the naming, we should use something more meaningful - like `fastFilter` or `passCellParams`.

Reply:
I also second the idea of moving in a path making migration pain lesser. +1 for attaching the v7 property to the function object.

Reply:
Alright, I'm convinced 👍

Reply:
https://github.com/mui/mui-x/pull/9254